framework/src/actions/AddPeriodicBCAction.C

```diff
@@ -146,66 +148,103 @@ AddPeriodicBCAction::autoTranslationBoundaries()
   {
-    RealVectorValue translation = getParam<RealVectorValue>("translation");
+    auto rm_params = _factory.getValidParams("ElementSideNeighborLayers");
 
-    PeriodicBoundary p(translation);
-    p.myboundary = _mesh->getBoundaryID(getParam<BoundaryName>("primary"));
-    p.pairedboundary = _mesh->getBoundaryID(getParam<BoundaryName>("secondary"));
-    setPeriodicVars(p, getParam<std::vector<VariableName>>("variable"));
+    rm_params.set<std::string>("for_whom") = "PeriodicBCs";
+    rm_params.set<MooseMesh *>("mesh") = Action::_mesh.get();
+    rm_params.set<Moose::RelationshipManagerType>("rm_type") =
+        Moose::RelationshipManagerType::GEOMETRIC | Moose::RelationshipManagerType::ALGEBRAIC;
 
-    _problem->addGhostedBoundary(p.myboundary);
-    _problem->addGhostedBoundary(p.pairedboundary);
+    // We can't attach this relationship manager early because
+    // the PeriodicBoundaries object needs to exist!
+    rm_params.set<bool>("attach_geometric_early") = false;
 
-    nl.dofMap().add_periodic_boundary(p);
-    if (displaced_problem)
-      displaced_problem->nlSys().dofMap().add_periodic_boundary(p);
-  }
-  else if (getParam<std::vector<std::string>>("transform_func") != std::vector<std::string>())
-  {
-    std::vector<std::string> inv_fn_names =
-        getParam<std::vector<std::string>>("inv_transform_func");
-    std::vector<std::string> fn_names = getParam<std::vector<std::string>>("transform_func");
-
-    // If the user provided a forward transformation, he must also provide an inverse -- we can't
-    // form the inverse of an arbitrary function automatically...
-    if (inv_fn_names == std::vector<std::string>())
-      mooseError("You must provide an inv_transform_func for FunctionPeriodicBoundary!");
-
-    FunctionPeriodicBoundary pb(*_problem, fn_names);
-    pb.myboundary = _mesh->getBoundaryID(getParam<BoundaryName>("primary"));
-    pb.pairedboundary = _mesh->getBoundaryID(getParam<BoundaryName>("secondary"));
-    setPeriodicVars(pb, getParam<std::vector<VariableName>>("variable"));
-
-    FunctionPeriodicBoundary ipb(*_problem, inv_fn_names);
-    ipb.myboundary = _mesh->getBoundaryID(getParam<BoundaryName>("secondary")); // these are swapped
-    ipb.pairedboundary =
-        _mesh->getBoundaryID(getParam<BoundaryName>("primary")); // these are swapped
-    setPeriodicVars(ipb, getParam<std::vector<VariableName>>("variable"));
-
-    _problem->addGhostedBoundary(ipb.myboundary);
-    _problem->addGhostedBoundary(ipb.pairedboundary);
-
-    // Add the pair of periodic boundaries to the dof map
-    nl.dofMap().add_periodic_boundary(pb, ipb);
-    if (displaced_problem)
-      displaced_problem->nlSys().dofMap().add_periodic_boundary(pb, ipb);
+    if (rm_params.areAllRequiredParamsValid())
+    {
+      auto rm_obj =
+          _factory.create<RelationshipManager>("ElementSideNeighborLayers",
+                                               "periodc_bc_ghosting" + Moose::stringify(count++),
```

Comment:
I'm not sure I love this `count` business. Why not use the `for_whom` field here instead? I don't think we need to add more than one of these per field.